File: Chroma-Core/src/main/java/buttondevteam/lib/chat/CoreArgumentCommandNode.java
package buttondevteam.lib.chat;

import com.mojang.brigadier.Command;
import com.mojang.brigadier.RedirectModifier;
import com.mojang.brigadier.arguments.ArgumentType;
import com.mojang.brigadier.builder.RequiredArgumentBuilder;
import com.mojang.brigadier.suggestion.SuggestionProvider;
import com.mojang.brigadier.tree.ArgumentCommandNode;
import com.mojang.brigadier.tree.CommandNode;

import java.util.function.Predicate;

public class CoreArgumentCommandNode<S, T> extends ArgumentCommandNode<S, T> {
	private final boolean optional;

	public CoreArgumentCommandNode(String name, ArgumentType<T> type, Command<S> command, Predicate<S> requirement, CommandNode<S> redirect, RedirectModifier<S> modifier, boolean forks, SuggestionProvider<S> customSuggestions, boolean optional) {
		super(name, type, command, requirement, redirect, modifier, forks, customSuggestions);
		this.optional = optional;
	}

	@Override
	public String getUsageText() {
		return optional ? "[" + getName() + "]" : "<" + getName() + ">";
	}

	@Override
	public RequiredArgumentBuilder<S, T> createBuilder() {
		return super.createBuilder();
	}
}

File: Chroma-Core/src/main/java/buttondevteam/lib/chat/CoreArgumentCommandNode.kt
package buttondevteam.lib.chat

import buttondevteam.lib.chat.commands.CommandUtils.coreArgument
import buttondevteam.lib.chat.commands.SubcommandData
import com.mojang.brigadier.Command
import com.mojang.brigadier.RedirectModifier
import com.mojang.brigadier.arguments.ArgumentType
import com.mojang.brigadier.builder.RequiredArgumentBuilder
import com.mojang.brigadier.suggestion.SuggestionProvider
import com.mojang.brigadier.tree.ArgumentCommandNode
import com.mojang.brigadier.tree.CommandNode
import java.util.function.Predicate

class CoreArgumentCommandNode<S : Command2Sender, T>(
    name: String?, type: ArgumentType<T>?, command: Command<S>?, requirement: Predicate<S>?, redirect: CommandNode<S>?, modifier: RedirectModifier<S>?, forks: Boolean, customSuggestions: SuggestionProvider<S>?,
    val optional: Boolean
) :
    ArgumentCommandNode<S, T>(name, type, command, requirement, redirect, modifier, forks, customSuggestions) {
    private var _commandData: SubcommandData<*, S>? = null
    var commandData: SubcommandData<*, S>
        get() {
            return _commandData
                ?: throw UninitializedPropertyAccessException("Command data has not been initialized")
        }
        internal set(value) {
            _commandData = value
            children.forEach { it.coreArgument()?.commandData = value }
        }

    override fun getUsageText(): String {
        return if (optional) "[$name]" else "<$name>"
    }

    override fun createBuilder(): RequiredArgumentBuilder<S, T> {
        return super.createBuilder()
    }
}